
# SCDE Analysis

> Kharchenko, Peter V., Lev Silberstein, and David T. Scadden. "Bayesian approach to single-cell differential expression analysis." Nature methods 11.7 (2014): 740-742.

## Fit Error Model

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}
if(is.null(r_data$scde_ifm)) return()
    DT::datatable(r_data$scde_ifm,options = list(scrollX = TRUE, scrollY = "210px", searching=F, order = list(list(5, 'asc')) , orderClasses = T))
```


## Probability of Transcript-detection Failures

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 8, fig.height = 6}
if(is.null(r_data$scde_ifm)) return()
    # get failure probabilities on the expresison range
    o.fail.curves <- scde.failure.probability(r_data$scde_ifm,magnitudes=log((10^r_data$scde_prior$x)-1))
    par(mfrow=c(1,1),mar = c(3.5,3.5,0.5,0.5), mgp = c(2.0,0.65,0), cex = 1);
    
    plot(c(),c(),xlim=range(r_data$scde_prior$x),ylim=c(0,1),xlab="expression magnitude (log10)",ylab="drop-out probability")
    
    sp <- r_data$sample_name
    nSamples <- length(sp)
    colorMin = 1
    colorMax = nSamples
    # colorRamp produces custom palettes, using values between 0-1
    colorFunction = colorRamp(c("black", "blue", "red"))
    zMin = colorMin
    zMax = colorMax
    zDiff = zMax - zMin
    z = zMin:zMax
    zScaled = (z - zMin) / zDiff
    # Apply colorRamp
    zMatrix = colorFunction(zScaled)
    # switch to hexadecimal representation and sort to be consistent with colorFactor.
    zColors = sort(rgb(zMatrix[,1], zMatrix[,2], zMatrix[,3], maxColorValue=255))
    
    # create an empty character vector that will be used to store the line colors. This is necessary to color the legend.
    colorsUsed = character(nSamples)

    for(i in 1:nSamples) {
        # using the sample name means the column order in colorFactorSorted file doesn't have to match the order in the counts file
        sample = sp[i]
        color = zColors[i]
        colorsUsed[i] = color
        lines(x=r_data$scde_prior$x,y=o.fail.curves[,sample], type="l", lty=i, lwd=1.5, cex=1, col=color)
    }
    legend("topright", legend=sp, col=colorsUsed, lty=1:nSamples, lwd=1.8, cex=0.8)
    
```

## Differential Expression Test result

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}

    ####### Delete this code chunk if you haven't run DE test, otherwise report generation will fail #######  

 if(is.null(r_data$scde_ediff)) return()
    
    if(!is.null(r_data$scde_batch)) { # If batch correction applied, it returns a list containing 3 dataframes rather than a single dataframe
        tbl <- r_data$scde_ediff[[input$scde_batch_ediff_choice]]
    } else {
        tbl <-r_data$scde_ediff
    }
    
    tbl <- tbl %>% dplyr::mutate(abs_Z = abs(tbl$Z))
    DT::datatable(tbl, selection = 'single', options = list(scrollX = TRUE, scrollY = "210px", searching=T, order = list(list(7, 'desc')), orderClasses = T))
        
```

## Plot of Selected Gene: `r input$scde_ediff_tbl_row_last_clicked`

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}

    ######## Delete this code chunk if you haven't selected any gene in the DE table, otherwise report generation will fail #######

if(is.null(r_data$scde_ediff)) return()

    s = input$scde_ediff_tbl_row_last_clicked
    if(is.null(s)) return({
        "No gene was selected. The plot will be random. Delete this code chunk if you don't need the gene plot."
    })
    
    if (length(s)) {
        if(!is.null(r_data$scde_batch)) {
            r_data$scde_gene <- rownames(r_data$scde_ediff$batch.adjusted[s, , drop = FALSE])
        }
        else {
            r_data$scde_gene <- rownames(r_data$scde_ediff[s, , drop = FALSE])
        }
    } else return()
    
    
    batch_show <- r_data$scde_batch # If no batch_info, this will be NULL. if batch info, then if the user choose batch.effect or batch.adjusted, show batch, otherwise set to NULL
    
    if(!is.null(r_data$scde_batch) && (input$scde_batch_ediff_choice == 'results')) {
        batch_show <- NULL
    }
    
    scde.test.gene.expression.difference(r_data$scde_gene, models = r_data$scde_ifm[r_data$scde_sample,], groups = r_data$scde_group, batch = batch_show, counts = r_data$raw, prior = r_data$scde_prior)
```
