
# Cell State Ordering by Monocle

> Cole Trapnell and Davide Cacchiarelli et al (2014): The dynamics and regulators of cell fate decisions are revealed by pseudo-temporal ordering of single cells. Nature Biotechnology

## State and Pseudotime Prediction

### Result Table

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}
    if(is.null(pData(r_data$cellset)$State)){
        return()
    }
    if(!is.null(r_data$group) && is.null(r_data$batch)) {
        tbl <- subset(pData(r_data$cellset), select = c(Pseudotime, State, Group))
    } else if(!is.null(r_data$group) && !is.null(r_data$batch)) {
        tbl <- subset(pData(r_data$cellset), select = c(Pseudotime, State, Group, Batch))
    } else if(is.null(r_data$group) && !is.null(r_data$batch)) {
        tbl <- subset(pData(r_data$cellset), select = c(Pseudotime, State, Batch))
    } else {
        
    }
    tbl <- pData(r_data$cellset)
    tbl <- tbl[, which(colnames(tbl)%in% c("Pseudotime", "State", "Group", "Batch"))]

    DT::datatable(tbl, options = list(
                      scrollX = T, scrollY = "400px",lengthMenu = c(20, 50, 100)
                  )
    )
```

### State and Group Comparison

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}
    if(is.null(pData(r_data$cellset)$Group) || is.null(pData(r_data$cellset)$State)) return()
    datatable(as.data.frame.matrix(table(pData(r_data$cellset)$Group, pData(r_data$cellset)$State)))
    plot(as.factor(pData(r_data$cellset)$Group), pData(r_data$cellset)$State, xlab="Group", ylab = "State")
```    

### Minimum Spanning Tree

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}
if(is.null(pData(r_data$cellset)$State)){
        return()
    }
    tree1 = ifelse(input$monocle_state_show %in% c("tree", "both"), T, F)
    mst1 = ifelse(input$monocle_state_show %in% c("mst", "both"), T, F)
    
    s = input$monocle_state_gene_tbl_row_last_clicked
    if(is.null(s) || length(s) == 0) {
        marker1 <- NULL
    }
    
    if(r_data$monocle_ordering == "de") {
        if(is.null(r_data$monocle_results)) return()
        tbl <- subset(r_data$monocle_results, qval <= 0.1) %>% tibble::rownames_to_column("gene")%>% dplyr::select(-pval)
        marker1 <- tbl[as.numeric(s),]$gene 
    } else {
        tbl <- data.frame(gene = rownames(fData(r_data$cellset)))
        marker1<- tbl[s,]
    }
    
    if(length(marker1) == 0) marker1 <- NULL
    plot_spanning_tree(r_data$cellset,  color_by = input$monocle_state_plt_color, show_tree = tree1, show_backbone = mst1, marker = marker1)
```    

## Gene Expression in Pseudotime

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}
        if(is.null(pData(r_data$cellset)$State) || is.null(r_data$monocle_genelist)){
        return()
    }
    cds_subset <- r_data$cellset[r_data$monocle_genelist, ]
    plot_genes_in_pseudotime(cds_subset, color_by = input$monocle_time_plt_color, cell_size = 3)
```  

## Cluster Genes by Psudotemporal Expression Pattern

### Result

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}
    if(is.null(r_data$monocle_clusters)) return()
    tbl <- data.frame(cluster = r_data$monocle_clusters$clustering)
    DT::datatable(tbl, 
                  options = list(
                      scrollX = T, scrollY = "400px", lengthMenu = c(20, 50, 100)
                  )
    )
``` 

### Plot
```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}
    if(is.null(r_data$monocle_clusters) || is.null(pData(r_data$cellset)$State)) return()
    plot_clusters(r_data$cellset, r_data$monocle_clusters)
``` 
